refactor(treesitter): use coroutines for resuming _parse() logic

ribru17 opened the pull request:
This means that all work previously done by a `_parse()` iteration will be kept in future iterations. This prevents it from running indefinitely in some cases where the file is very large and there are 2+ injections.

This PR also allows `is_valid()` to accept a range, which narrows the area for which to check the tree for validity. This fixes #32075 with the downside that the `_valid` table can no longer be compressed to a `true` value. However, I would argue that this loss is very minimal because the compression happens almost never for large files (which is where it actually needs to happen)

close #32050 
close #32077

ribru17 committed:
refactor(treesitter): use coroutines for resuming _parse() logic

This means that all work previously done by a `_parse()` iteration will
be kept in future iterations. This prevents it from running indefinitely
in some cases where the file is very large and there are 2+ injections.

vanaigr commented on runtime/lua/vim/treesitter/languagetree.lua:
```diff
@@ -484,7 +497,8 @@ function LanguageTree:_async_parse(range, on_parse)
       total_parse_time = 0
```
I think the coroutine should also be recreated

vanaigr commented:
It still runs infinitely, but now it is because the async parse doesn't finish in 1 step, causes a redraw, which schedules another async parse which also doesn't finish and causes another redraw

ribru17 commented:
Perfect. That will be addressed by #32077 (which needs to be amended a bit after this PR)

tomtomjhj commented:
I ran into a type error on `self._valid[i] = true` in `LanguageTree:_parse_regions()`.

This happened when
* treesitter highlighting and foldexpr is active,
* and when I did lsp go to definition, which opens a big lua file.

Unfortunately, I couldn't reproduce this with a simple setup, but here is brief description of what happened:

1.  Call `parse(nil, function() ... end)`.
    (In my case, this was invoked by the initial evaluation of `foldexpr`.)
    At this point, `changedtick` was 1.
    `parse` does not finish in the given time frame, so it yields, and schedules the next step.
2.  (Without passing the control to user) `changedtick` is changed to 2, and `parse` is invoked again (separate from the previous invocation).
    (This time, it was invoked by treesitter highlighter's `on_win` callback.)
    Again, the parsing does not finish in the given time frame, so it yields, and schedules the next step.
3.  The scheduled step with `ct = 1` is resumed.
    Since `changedtick` is now 2, it resets the parsing state, and restarts `parse(nil)` from the beginning.
    In `_parse()`, this calls `self:is_valid(true)`,
    At this point, `self._valid` was `{ true }`, so it is compressed to `true`.
4.  The scheduled step with `ct = 2` is resumed.
    In `_parse_regions()`, it attempts to set `self._valid[i] = true`, but `self._valid` is now a boolean, so it raises an error.

In summary, the problem is that async `parse` can be invoked multiple times with different `changedtick`, and the LanguageTree state is shared across these invocations.
Currently, TSHighlighter prevents this with `self.parsing` flag, but there still are cases where another party invokes async parse.

ribru17 commented:
Thanks for the detailed explanation, I'll investigate this now. I am curious why the changedtick gets incremented even if no edit was made?

ribru17 commented:
@tomtomjhj I've pulled in #32077 and I think it fixes the issue here. Can you confirm that this is the case? The main fix is just the fact that `._valid` is no longer compressed to a single `true` value.

lewis6991 commented:
> The main fix is just the fact that ._valid is no longer compressed to a single true value.

I haven't followed the issue closely, but can you explain the rationale for this? If a tree is valid, there should be no need to keep scanning the valid table.

ribru17 commented:
I have found that the table rarely gets compressed to `true` in situations where it is actually needed. This only happens when all regions have been parsed, which is only the case if the user has scrolled each region into the window viewport (if we are talking about the highlighter, of course the user can manually run `:parse(true)` which would parse everything). With the change of allowing `is_valid()` to accept a range, I think the table compressing adds too much complexity (and apparently some bugginess) to justify its gain, since it only really optimizes the case for when `is_valid()` is not given a range, which will not be the case in the highlighter. Also we never iterate `_valid` directly, only indexing it after iterating `included_regions`.

tomtomjhj reviewed (commented):
The new commit fixes the issue i'm experiencing.

lewis6991 commented:
> have found that the table rarely gets compressed to true in situations where it is actually needed. This only happens when all regions have been parsed, which is only the case if the user has scrolled each region into the window viewport (if we are talking about the highlighter, of course the user can manually run :parse(true) which would parse everything).

Well yes, the highlighter is just one application with its own requirements. Other plugins may need to parse the whole tree.

> With the change of allowing is_valid() to accept a range, I think the table compressing adds too much complexity (and apparently some bugginess) to justify its gain, 

Did it really add that much complexity? From memory all the logic was contained in `is_valid()`, and it was only set to the boolean at the end, and checked at the start. This is a pretty common idiom to cache results of a function.

> since it only really optimizes the case for when is_valid() is not given a range, which will not be the case in the highlighter. 

Again, highlighter is just one application. We're not optimizing purely for that.

> Also we never iterate _valid directly, only indexing it after iterating included_regions.

Sorry, I don't quite understand this.

ribru17 commented:
> Did it really add that much complexity? From memory all the logic was contained in `is_valid()`, and it was only set to the boolean at the end, and checked at the start. This is a pretty common idiom to cache results of a function.

Sure, in this case maybe I can keep that logic there and then only compress the value in `is_valid()` when no range parameter is present. In this case, do you have another idea of how to solve the issue from above? Maybe return early from a `parse_regions()` iteration if `_valid` has become `true`

> > Also we never iterate _valid directly, only indexing it after iterating included_regions.
> 
> Sorry, I don't quite understand this.

Basically, it is never the case that we iterate over the values in `_valid` (when it is a table). We always only iterate over the included regions, at that point checking if a corresponding entry exists in `_valid`. I just mentioned that to note that having a large `_valid` table doesn't really hurt performance too much on it's own, although on the other hand it means that there are many included regions, meaning iterating over those can be slow. But the point is moot if we just keep the compression logic

ribru17 commented:
Another idea: move `_valid` to a `_valid_regions` member variable, and keep another `.valid` variable as a simple boolean, only set to true when all regions are valid (akin to compressing the original `_valid`), and reset to `false` as needed in `iter_regions`

ribru17 committed:
feat(treesitter): allow LanguageTree:is_valid() to accept a range

When given, only that range will be checked for validity rather than the
entire tree. This is used in the highlighter to save CPU cycles since we
only need to parse a certain region at a time anyway.

ribru17 commented:
Some changes to note:

- Instead of compressing `_valid` to `true`, we now have `_is_entirely_valid` to show that the tree is wholly valid
- `_parse_regions()` now sets `_is_entirely_valid` to `true` when it has parsed the last invalid region
- `_valid` is now `_valid_regions` which is always a table; more specifically, a set of valid regions
- Unlike `_valid`, this table can only hold `true` values. `false` values are instead set to `nil`. This fixes a bug I discovered that existed before async parsing, in which `set_included_regions` would incorrectly think that all regions were valid because the size of `_valid` was equal to the number of included regions (https://github.com/neovim/neovim/blob/35c5e231078365033524b0aa2166118a1b2ef600/runtime/lua/vim/treesitter/languagetree.lua#L755). This was previously not correct because `false` values would also contribute to the count. This has been fixed, which is also what allows `_parse_regions()` to update the validity state upon parsing, simplifying the logic in `is_valid()` (and removing the side effects in `is_valid()`)

tomtomjhj commented on runtime/lua/vim/treesitter/languagetree.lua:
```diff
@@ -320,64 +350,42 @@ function LanguageTree:source()
       local parse_time, tree, tree_changes =
         tcall(self._parser.parse, self._parser, self._trees[i], self._source, true)
+      while true do
+        if not tree then
+          coroutine.yield(changes, no_regions_parsed, total_parse_time, false)
+        else
+          break
+        end
```
```suggestion
        if tree then
          break
        end
        coroutine.yield(changes, no_regions_parsed, total_parse_time, false)
```

tomtomjhj commented on runtime/lua/vim/treesitter/languagetree.lua:
```diff
@@ -558,12 +576,20 @@ function LanguageTree:_parse(range, timeout)
+        parse_regions(self, range, thread_state)
+      thread_state.timeout = thread_state.timeout
+        and math.max(thread_state.timeout - total_parse_time, 0)
+      if not is_finished then
+        coroutine.yield(self._trees, false)
+      else
+        break
+      end
```
```suggestion
      if is_finished then
        break
      end
      coroutine.yield(self._trees, false)
```

tomtomjhj commented on runtime/lua/vim/treesitter/languagetree.lua:
```diff
@@ -389,7 +397,11 @@ function LanguageTree:_parse_regions(range, timeout)
-      self._valid[i] = true
+      self._valid_regions[i] = true
+
+      if #self._valid_regions == #self:included_regions() then
+        self._is_entirely_valid = true
+      end
```
What happens if `_valid_regions` has holes?

ribru17 replied:
Ah, good point. This wouldn't matter if `included_regions()` didn't have holes, but since it does it looks like a different solution is needed

ribru17 replied:
Let me know what you think of the new solution. I have opted to store the number of included regions during `set_included_regions()`, which is the only time when there can be no holes and thus the length is a true representation of the number of regions (to my understanding)

tomtomjhj replied:
Overall idea looks good to me, but there are potentially missing updates for `_num_regions` at
* https://github.com/neovim/neovim/blob/e8de4338be74a6817d21c9d707cebb2edb42233e/runtime/lua/vim/treesitter/languagetree.lua#L828
* https://github.com/neovim/neovim/blob/e8de4338be74a6817d21c9d707cebb2edb42233e/runtime/lua/vim/treesitter/languagetree.lua#L1080

ribru17 replied:
This was intentional, but let me know if there is a misunderstanding. These two locations aren't really changing the total amount of regions, rather just invalidating them (second example) or repopulating them, but with holes (first example). I believe the count calculation should stay only in `set_included_regions()` because this is the only place that actually sets the full list of regions, which is what the count should represent. If the number of regions will actually change after setting `_regions = nil`,  then that is also file because it will be updated after another call to `set_included_regions()`

tomtomjhj replied:
aha that makes sense

tomtomjhj commented on runtime/lua/vim/treesitter/languagetree.lua:
```diff
@@ -480,7 +492,11 @@ function LanguageTree:_async_parse(range, on_parse)
   local redrawtime = vim.o.redrawtime
-  local timeout = not vim.g._ts_force_sync_parsing and default_parse_timeout_ms or nil
+
+  local thread_state = {} ---@type ParserThreadState
```
Why does this need to be a table? If it should be a table, I think it should also include `ct`, because it's also a mutable state used in `ct`.

tomtomjhj replied:
I just thought it would look more organized if all the mutable thread states are collected into a single table (including `total_parse_time`, which I missed in the previous comment). But in terms of functionality there is no difference. Please feel free to disregard.

tomtomjhj reviewed (commented):
> This fixes a bug I discovered that existed before async parsing, in which `set_included_regions` would incorrectly think that all regions were valid because the size of `_valid` was equal to the number of included regions. This was previously not correct because false values would also contribute to the count.

But `_iter_regions` checks `_valid[i]`. If `false`, then `all_valid = false`. So that problem does not happen.

https://github.com/neovim/neovim/blob/35c5e231078365033524b0aa2166118a1b2ef600/runtime/lua/vim/treesitter/languagetree.lua#L703-L716

ribru17 commented:
> But `_iter_regions` checks `_valid[i]`. If `false`, then `all_valid = false`. So that problem does not happen.

You're right, I misread that part of the code :+1:

justinmk commented on runtime/lua/vim/treesitter/languagetree.lua:
```diff
@@ -271,16 +279,46 @@ function LanguageTree:lang()
+    if not range then
+      return false
+    end
+    -- TODO: Efficiently search for possibly intersecting regions using a binary search
```
Related idea in our internal `marktree`: https://github.com/neovim/neovim/blob/af069c5c05ad99623345071007ad23da51973601/src/nvim/marktree.c#L331-L332

justinmk reviewed (approved):
Anyone want more time to review this? Else we should just merge it to let it "bake" before 0.11
